bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/Tr064RootHandler.java

```diff
@@ -0,0 +1,391 @@
+                }
+            }
+        });
+    }
+
+    /**
+     * establish the connection - get secure port (if avallable), install authentication, get device properties
+     *
+     * @return true if successful
+     */
+    private boolean establishSecureConnectionAndUpdateProperties() {
+        final SCPDUtil scpdUtil = this.scpdUtil;
+        if (scpdUtil != null) {
+            try {
+                SCPDDeviceType device = scpdUtil.getDevice("")
+                        .orElseThrow(() -> new SCPDException("Root device not found"));
+                SCPDServiceType deviceService = device.getServiceList().stream()
+                        .filter(service -> service.getServiceId().equals("urn:DeviceInfo-com:serviceId:DeviceInfo1"))
+                        .findFirst().orElseThrow(() -> new SCPDException(
+                                "service 'urn:DeviceInfo-com:serviceId:DeviceInfo1' not found"));
+
+                this.deviceType = device.getDeviceType();
+
+                // try to get security (https) port
+                SOAPMessage soapResponse = soapConnector.doSOAPRequest(deviceService, "GetSecurityPort",
+                        Collections.emptyMap());
+                if (!soapResponse.getSOAPBody().hasFault()) {
+                    SOAPValueConverter soapValueConverter = new SOAPValueConverter(httpClient);
+                    soapValueConverter.getStateFromSOAPValue(soapResponse, "NewSecurityPort", null)
+                            .ifPresentOrElse(port -> {
+                                endpointBaseURL = "https://" + config.host + ":" + port.toString();
+                                soapConnector = new SOAPConnector(httpClient, endpointBaseURL);
+                                logger.debug("endpointBaseURL is now '{}'", endpointBaseURL);
+                            }, () -> logger.warn("Could not determine secure port, disabling https"));
+                } else {
+                    logger.warn("Could not determine secure port, disabling https");
+                }
+
+                // clear auth cache and force re-auth
+                httpClient.getAuthenticationStore().clearAuthenticationResults();
+                AuthenticationStore auth = httpClient.getAuthenticationStore();
+                auth.addAuthentication(new DigestAuthentication(new URI(endpointBaseURL), Authentication.ANY_REALM,
+                        config.user, config.password));
+
+                // check & update properties
+                SCPDActionType getInfoAction = scpdUtil.getService(deviceService.getServiceId())
+                        .orElseThrow(() -> new SCPDException(
+                                "Could not get service definition for 'urn:DeviceInfo-com:serviceId:DeviceInfo1'"))
+                        .getActionList().stream().filter(action -> action.getName().equals("GetInfo")).findFirst()
+                        .orElseThrow(() -> new SCPDException("Action 'GetInfo' not found"));
```

Comment:
I wonder if it is worth it to make utility methods for these frequently used name lookups...

```
private Function<SCPDActionType, Optional<SCPDArgumentType> getArgument(String argumentName, String direction){
    return (action) -> action.getArgumentList().stream()
                                        .filter(argument -> argument.getName().equals(argumentName)
                                                && argument.getDirection().equals(direction))
                                        .findFirst();
}

private Function<SCPDServiceType,  Optional<SCPDActionType>> getAction(String actionName){
    return (service) -> service.getActionList().stream().filter(action -> action.getName().equals(actionName)).findFirst();
}
```
```suggestion
                SCPDActionType getInfoAction = scpdUtil.getService(deviceService.getServiceId())
                        .orElseThrow(() -> new SCPDException(
                                "Could not get service definition for 'urn:DeviceInfo-com:serviceId:DeviceInfo1'"))
                        .flatMap(getAction("GetInfo"))
                        .orElseThrow(() -> new SCPDException("Action 'GetInfo' not found"));
```

Reply:
Where do you see a benefit here? It's only used in this place AFAIK.

Reply:
It is used in several places in the Util class. But frankly I'm not sure my suggestion provides enough benefit to be worth the time to do it anyway.